Cleanup of `ByteBufferAccumulator` (approach 2)

gregw opened the pull request:
Alternative to #11058 to fix #10541

gregw committed:
Review ByteBufferAccumulator #10541

 + Added `isRetained()` to `Retainable`
 + Made `Chunk` is-a `RetainableByteBuffer`
 + retain rather than copy chunks where possible

gregw committed:
Review ByteBufferAccumulator #10541

 + Added `isRetained()` to `Retainable`
 + Made `Chunk` is-a `RetainableByteBuffer`
 + retain rather than copy chunks where possible

gregw committed:
Review ByteBufferAccumulator #10541

 + Added `isRetained()` to `Retainable`
 + Made `Chunk` is-a `RetainableByteBuffer`
 + retain rather than copy chunks where possible

gregw committed:
Review ByteBufferAccumulator #10541

 + Added `isRetained()` to `Retainable`
 + Made `Chunk` is-a `RetainableByteBuffer`
 + retain rather than copy chunks where possible
 + Merged aggregator into accumulator
 + Deleted ByteBufferAggregator.java

gregw committed:
Review ByteBufferAccumulator #10541

 + Added `isRetained()` to `Retainable`
 + Made `Chunk` is-a `RetainableByteBuffer`
 + retain rather than copy chunks where possible
 + Merged aggregator into accumulator
 + Deleted ByteBufferAggregator.java

gregw committed:
Merge branch 'jetty-12.0.x' into fix/jetty-12/10541/byteBufferAccumulator

gregw committed:
Review ByteBufferAccumulator #10541

 + Added `isRetained()` to `Retainable`
 + Made `Chunk` is-a `RetainableByteBuffer`
 + retain rather than copy chunks where possible
 + Merged aggregator into accumulator
 + Deleted ByteBufferAggregator.java

gregw committed:
Review ByteBufferAccumulator #10541

 + Added `isRetained()` to `Retainable`
 + Made `Chunk` is-a `RetainableByteBuffer`
 + retain rather than copy chunks where possible
 + Merged aggregator into accumulator
 + Deleted ByteBufferAggregator.java

gregw committed:
Review ByteBufferAccumulator #10541

 + Added `isRetained()` to `Retainable`
 + Made `Chunk` is-a `RetainableByteBuffer`
 + retain rather than copy chunks where possible
 + Merged aggregator into accumulator
 + Deprecated ByteBufferAggregator.java
 + Deprecated ByteBufferCallbackAccumulator

gregw committed:
Review ByteBufferAccumulator #10541

 + Added `isRetained()` to `Retainable`
 + Made `Chunk` is-a `RetainableByteBuffer`
 + retain rather than copy chunks where possible
 + Merged aggregator into accumulator
 + Deprecated ByteBufferAggregator.java
 + Deprecated ByteBufferCallbackAccumulator

gregw committed:
Merge branch 'jetty-12.0.x' into fix/jetty-12/10541/byteBufferAccumulator

gregw committed:
Review ByteBufferAccumulator #10541

 + Added append and writeTo methods to RetainableByteBuffer
 + Made Aggregator and Accumulator subtypes of RetainableByteBuffer

gregw committed:
Review ByteBufferAccumulator #10541

 + Added append and writeTo methods to RetainableByteBuffer
 + Made Aggregator and Accumulator subtypes of RetainableByteBuffer

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:
```diff
@@ -159,6 +170,403 @@ default void clear()
+        private final boolean _direct;
+        private final long _maxLength;
+        private final List<RetainableByteBuffer> _buffers = new ArrayList<>();
+        private boolean _lastAcquiredByUs;
```
I think all the logic using that boolean is wrong. See other comments.

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:
```diff
@@ -159,6 +170,403 @@ default void clear()
+            if (length < remaining)
+                return false;
+
+            _buffers.add(new RetainableByteBuffer()
```
This `RetainableByteBuffer` impl breaks the retain/release contract. What I think should be done instead is to create a new `wrap` static helper in `RetainableByteBuffer` as follows:

```java
    static RetainableByteBuffer wrap(ByteBuffer byteBuffer, Runnable releaser)
    {
        return new AbstractRetainableByteBuffer(byteBuffer) {
            @Override
            public boolean release()
            {
                boolean released = super.release();
                if (released)
                    releaser.run();
                return released;
            }
        };
    }
```
then all contained RBBs can be released on `clear`.

gregw committed:
Merge branch 'jetty-12.0.x' into fix/jetty-12/10541/byteBufferAccumulator2

gregw committed:
Merge branch 'jetty-12.0.x' into fix/jetty-12/10541/byteBufferAccumulator

gregw committed:
Merge branch 'fix/jetty-12/10541/byteBufferAccumulator' into fix/jetty-12/10541/byteBufferAccumulator2

gregw committed:
Review ByteBufferAccumulator #10541

 + Added append and writeTo methods to RetainableByteBuffer
 + Made Aggregator and Accumulator subtypes of RetainableByteBuffer

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/AsyncContent.java:
```diff
@@ -301,6 +301,12 @@ public boolean canRetain()
+        @Override
+        public boolean isRetained()
+        {
+            return referenceCounter.isRetained();
```
```diff
--- return referenceCounter.isRetained();
+++ return referenceCounter != null && referenceCounter.isRetained();
```

lorban commented on jetty-core/jetty-io/src/test/java/org/eclipse/jetty/io/RetainableByteBufferTest.java:
```diff
@@ -0,0 +1,118 @@
+    public static void beforeAll()
+    {
+        _pool = new ArrayByteBufferPool.Tracking(MIN_CAPACITY, MIN_CAPACITY, MAX_CAPACITY, Integer.MAX_VALUE);
+    }
```
You should add a leak check since you're using a `Tracking` pool:

```java
@AfterAll
public static void afterAll()
{
    assertThat("Leaks: " + bufferPool.dumpLeaks(), bufferPool.getLeaks().size(), is(0));
}
```

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:
```diff
@@ -151,6 +152,16 @@ default int capacity()
+
+    default boolean isFull()
+    {
+        return remaining() == capacity();
```
Shouldn't that be `return remaining() == 0`?

gregw replied:
No, as we are in flush mode and remaining is the size of valid bytes, not available empty bytes.

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:
```diff
@@ -159,6 +170,419 @@ default void clear()
         BufferUtil.clear(getByteBuffer());
     }
 
+    default int append(byte[] bytes, int offset, int length)
```
I'd rather not have variants of `append` working on `byte[]` unless necessary, and so far, only tests use this variant.

gregw replied:
I think it is worth while if we ever want this API to replace much of BufferUtil.  In fact I could probably find usages for this now by replace BufferUtil calls.

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:
```diff
@@ -159,6 +170,419 @@ default void clear()
+        return length;
+    }
+
+    default void append(ByteBuffer bytes)
```
Why not return `boolean` to indicate all bytes were copied?

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:
```diff
@@ -159,6 +170,419 @@ default void clear()
+        BufferUtil.append(getByteBuffer(), bytes);
+    }
+
+    default boolean append(ByteBuffer bytes, Runnable releaser)
```
This variant taking a `Runnable releaser` seems at odds with the release doctrine. I'm not sure I like it.

gregw replied:
I have removed it.  There is a wrapper in RetainableByteBuffer to make one from a releaser.  I agree it is strange, so that keeps the strangeness in one spot.

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:
```diff
@@ -159,6 +170,419 @@ default void clear()
+        return true;
+    }
+
+    default void append(RetainableByteBuffer bytes)
```
Why not return `boolean` to indicate all bytes were copied?

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:
```diff
@@ -159,6 +170,419 @@ default void clear()
+        bytes.writeTo(this);
+    }
+
+    default void putTo(ByteBuffer toInfillMode)
```
The difference between `putTo` and `writeTo` feels... error-prone. I'd rather not have such a helper unless it clearly helps avoid bugs.

gregw replied:
I have removed the writeTo to a buffer.  You can only write to a stream of some kind now.

sbordet replied:
What the parameter name even mean?
What's "Infill"?
The parameter is the buffer to write to, and this operation in the JDK is called `ByteBuffer.get()`.

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:
```diff
@@ -159,6 +170,419 @@ default void clear()
+
+    default boolean writeTo(ByteBuffer to)
+    {
+        ByteBuffer slice = getByteBuffer().slice();
```
Why not "consuming" the buffer instead of slicing it?

gregw replied:
I agree.  We should have all these methods consume.  If  you don't want to consume, then the caller can slice.
However, we may need a slice method on the RetainableByteBuffer API... then it gets a little confusing if that should be a retain/release or not.

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:
```diff
@@ -159,6 +170,419 @@ default void clear()
+        return slice.remaining() == 0;
+    }
+
+    default void writeTo(OutputStream out) throws IOException
```
I'd rather not have a variant that writes to an `OutputStream` unless needed, and so far, only tests use this method.

gregw replied:
The whole point is so that we can efficiently write accumulated buffers to something like an OutputStream.  I think it has to be there, since when it is needed an efficient version is unlikely to be written and getByteBuffer will just be called.

lorban replied:
Wouldn't a `Content.Sink` to `OutputStream` adapter like the following give the same benefits without needed extra API?

```java
public class OutputStreamContentSink implements Content.Sink
{
    private final OutputStream outputStream;
    private final int tempBufferSize;
    private final ByteBufferPool pool;
    private RetainableByteBuffer tempBuffer;
    private boolean lastWritten;

    public OutputStreamContentSink(OutputStream outputStream)
    {
        this(outputStream, 1024, new ByteBufferPool.NonPooling());
    }

    public OutputStreamContentSink(OutputStream outputStream, int tempBufferSize, ByteBufferPool pool)
    {
        this.outputStream = outputStream;
        this.tempBufferSize = tempBufferSize;
        this.pool = pool;
    }

    @Override
    public void write(boolean last, ByteBuffer byteBuffer, Callback callback)
    {
        if (lastWritten)
        {
            callback.failed(new IOException("complete"));
            return;
        }
        lastWritten = last;

        Throwable failure = null;
        try
        {
            if (byteBuffer.hasArray())
            {
                outputStream.write(byteBuffer.array(), byteBuffer.arrayOffset() + byteBuffer.position(), byteBuffer.remaining());
                byteBuffer.position(byteBuffer.position() + byteBuffer.remaining());
            }
            else
            {
                if (tempBuffer == null)
                {
                    tempBuffer = pool.acquire(tempBufferSize, false);
                }
                while (byteBuffer.hasRemaining())
                {
                    byte[] tempArray = tempBuffer.getByteBuffer().array();
                    int byteCountToWrite = Math.min(byteBuffer.remaining(), tempArray.length);
                    byteBuffer.get(tempArray, 0, byteCountToWrite);
                    outputStream.write(tempArray, 0, byteCountToWrite);
                }
            }
            callback.succeeded();
        }
        catch (Throwable x)
        {
            failure = x;
        }

        try
        {
            if (last)
            {
                if (tempBuffer != null)
                {
                    tempBuffer.release();
                    tempBuffer = null;
                }
                outputStream.close();
            }
        }
        catch (Throwable x)
        {
            if (failure == null)
                failure = x;
            else
                ExceptionUtil.addSuppressedIfNotAssociated(failure, x);
        }

        if (failure != null)
        {
            if (tempBuffer != null)
            {
                tempBuffer.release();
                tempBuffer = null;
            }
            callback.failed(failure);
        }
    }
}
```

gregw replied:
Hmmmm the PR has 1 new method, with a 1 line default implementation and a 2 line optimized implementation for aggregation.

This is a whole new class with a over 60 lines of impl (although it could be shortened to about 30 I think if it used BufferUtil.writeTo).

I'm not opposed to having such a class... but I just don't see it as simpler than the writeTo method.   Let me have a go of writing the simplest impl I can and then we can decide....

gregw replied:
OK, I was able to implement that with an 11 line anonymous lambda.   So I have removed this writeTo method.

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:
```diff
@@ -159,6 +170,419 @@ default void clear()
+
+        private Aggregator(ByteBufferPool pool, boolean direct, int growBy, int maxCapacity)
+        {
+            if (growBy < 0)
```
```diff
--- if (growBy < 0)
+++ if (growBy <= 0)
```

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:
```diff
@@ -159,6 +170,419 @@ default void clear()
+            if (growBy < 0)
+                throw new IllegalArgumentException("growBy must be > 0");
+            if (maxCapacity > 0 && growBy > maxCapacity)
+                throw new IllegalArgumentException("growBy must be < maxSize");
```
```diff
--- throw new IllegalArgumentException("growBy must be < maxSize");
+++ throw new IllegalArgumentException("growBy must be <= maxSize");
```

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:
```diff
@@ -159,6 +170,419 @@ default void clear()
+        @Override
+        public int capacity()
+        {
+            return Math.max(_buffer.capacity(), _maxCapacity);
```
I think this should be `Math.min()` as I don't think we want to overshoot `_maxCapacity`, even if the buffer is bigger than it.

gregw replied:
The style of interaction with RetainableByteBuffer has always been that you can bypass the API by getting the buffer directly and manipulating it.  So any capacity limit we put in this method can by bypassed anyway.
So I think it is better to embrace the fact that the pool might give us a bigger buffer than was requested and to use that extra capacity if already allocated.  I'll javadoc that.

gregw replied:
Alternately we could have an extra constructor that takes an additional `strictCapacity` boolean, that if true would always slice a buffer acquired from the pool to be exactly the desired capacity.

Or should we just have a variant of the pool that does that?

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:
```diff
@@ -159,6 +170,419 @@ default void clear()
+        }
+
+        @Override
+        public int append(byte[] bytes, int offset, int length)
```
I don't like this variant with `byte[]`.

gregw replied:
I guess we could always do the wrap externally, but it feels like a reasonable convenience method?

lorban replied:
Let's add it when we have a need for this method?

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:
```diff
@@ -159,6 +170,419 @@ default void clear()
+        }
+
+        @Override
+        public void append(ByteBuffer bytes)
```
I think this should return a `boolean`.

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:
```diff
@@ -159,6 +170,419 @@ default void clear()
+        }
+
+        @Override
+        public boolean append(ByteBuffer bytes, Runnable releaser)
```
I don't like this variant taking a `Runnable` releaser.

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:
```diff
@@ -159,6 +170,419 @@ default void clear()
+        }
+
+        @Override
+        public void append(RetainableByteBuffer bytes)
```
I think this should return a `boolean`.

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:
```diff
@@ -159,6 +170,419 @@ default void clear()
+        sink.write(last, getByteBuffer(), callback);
+    }
+
+    static RetainableByteBuffer newAggregator(ByteBufferPool pool, boolean direct, int growBy, int maxCapacity)
```
What is this factory method buying us over directly instantiating the `Aggregator` class?

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:
```diff
@@ -159,6 +170,419 @@ default void clear()
+     * @param maxLength The maximum length of the accumulated buffers or -1 for no limit
+     * @return The accumulating {@link RetainableByteBuffer}
+     */
+    static RetainableByteBuffer newAccumulator(ByteBufferPool bufferPool, boolean direct, int maxLength)
```
Why this factory method?

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:
```diff
@@ -159,6 +170,419 @@ default void clear()
+        }
+
+        @Override
+        public int remaining()
```
I'm worried about the long-to-int conversion. We should carefully weight and document if we throw or cap on int overflow.

lorban replied:
I'm wondering if we shouldn't cap the accumulator to 2GB?

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:
```diff
@@ -159,6 +170,419 @@ default void clear()
+        }
+
+        @Override
+        public int capacity()
```
Ditto.

gregw replied:
ditto

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:
```diff
@@ -159,6 +170,419 @@ default void clear()
+        }
+
+        @Override
+        public int append(byte[] bytes, int offset, int length)
```
I don't like this variant taking a `byte[]`.

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:
```diff
@@ -159,6 +170,419 @@ default void clear()
+        }
+
+        @Override
+        public void append(ByteBuffer bytes)
```
I think this should return `boolean`.

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/BufferedContentSink.java:
```diff
@@ -43,15 +40,12 @@ public class BufferedContentSink implements Content.Sink
     private final int _maxAggregationSize;
-    private final Flusher _flusher;
-    private ByteBufferAggregator _aggregator;
+    private RetainableByteBuffer _accumulator;
```
I think an `Aggregator` is the correct implementation here.

lorban commented:
I do like the new `Aggregator` and `Accumulator` design, including the fact that they are `RetainableByteBuffer` implementations.

But the API should be discussed as I would like it to be rock-solid as it's very central to a lot of potential problems.

gregw committed:
Review ByteBufferAccumulator #10541

WIP on review feedback

gregw commented:
@lorban I've acted on most of your feedback, but a couple of questions on some of it.
In general, I happy to reduce the API to the minimal to ensure we get it right... and then we need to document it well.

However, I'm only partially available over the next few weeks, so feel free to contribute to this branch.  I don't totally agree with the remaining API changes you suggest... but I don't totally disagree either, so he who does the work....
Also I think `BufferedContentSink` needs a lot of review at least, if not bug fixes.

gregw committed:
New implementation of BufferedContentSink

gregw committed:
javadoc

gregw commented on jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/ResponseHeadersTest.java:
```diff
@@ -150,6 +151,44 @@ protected void doGet(HttpServletRequest req, HttpServletResponse response) throw
     }
 
+    @Test
+    public void testCharsetAfterGetOutputStream() throws Exception
```
oops this was an accidental inclusion into this PR.   It is just a test to check existing behaviour.... I can remove if it irks people

gregw committed:
do not clear on release

gregw committed:
More fixes to BufferedContentSink

gregw committed:
forced flush

gregw committed:
Fixed bug in BufferUtil

gregw committed:
document long to int conversion

gregw committed:
Merge branch 'jetty-12.0.x' into fix/jetty-12/10541/byteBufferAccumulator2

gregw commented:
@sbordet @lorban I've punted this PR to 12.0.8, as it is a biggish change and needs time to settle before a release.  Can you both look at it early in the next release cycle (or now), so it will be ready in time.

gregw committed:
Merge branch 'jetty-12.0.x' into fix/jetty-12/10541/byteBufferAccumulator2

gregw commented:
@sbordet @lorban nudge

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/AsyncContent.java:
```diff
@@ -330,5 +336,17 @@ public void failed(Throwable x)
+            return "%s@%x[rc=%d,l=%b,b=%s]".formatted(
+                getClass().getSimpleName(),
+                hashCode(),
+                referenceCounter.get(),
```
`referenceCounter` can be null

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:
```diff
@@ -159,6 +192,433 @@ default void clear()
+         * {@link RetainableByteBuffer}s with zero-copy if the {@link #append(RetainableByteBuffer)} API is used
+         * @param pool The pool from which to allocate buffers
+         * @param direct true if direct buffers should be used
+         * @param growBy the size to grow the buffer by or &lt;= 0 for a heuristic
```
I have a feeling that `growBy` and `maxCapacity` should not have heuristics nor defaults, as I think those would need to be specific depending on what uses the aggregator.

gregw replied:
I agree the heuristic is not the best.   But it is just as hard for the caller to know the best value.  It really should be the buffer pools bucket increment.    Let me see if I can at least come up with a better heuristic.

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:
```diff
@@ -159,6 +192,433 @@ default void clear()
+                    {
+                        if (i < _buffers.size())
+                        {
+                            _buffers.get(i).writeTo(sink, ++i == _buffers.size() || !last, this);
```
We should never pass `true` as the 2nd argument if `last` is `false`.

gregw committed:
removed byte[] append API

gregw committed:
removed writeTo(OutputStream)

gregw committed:
updates from review

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java:
```diff
@@ -470,6 +470,29 @@ default boolean rewind()
+         * @param out The stream to wrap
+         * @return A sink wrapping the stream
+         */
+        static Sink from(OutputStream out)
```
Nice!

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ByteBufferAccumulator.java:
```diff
@@ -18,44 +18,53 @@
- *
- * The method {@link #ensureBuffer(int, int)} is used to write directly to the last buffer stored in the buffer list,
+ * <p>
+ * The method {@link #accessInternalBuffer(int, int)} can be used access a buffer that can be written to directly as the last buffer list,
```
I do not understand the following sentence:
"that can be written to directly _as the last buffer list_"

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ByteBufferAccumulator.java:
```diff
@@ -18,44 +18,53 @@
+ * @deprecated use {@link RetainableByteBuffer.Accumulator}
  */
-// TODO: rename to *Aggregator to avoid confusion with RBBP.Accumulator?
+@Deprecated
```
Why modify this class if it has to be deprecated?

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:
```diff
@@ -159,6 +190,370 @@ default void clear()
+            int length = ensureMaxLength(remaining);
+
+            // If we cannot retain, then try aggregation into the last buffer
+            if (!bytes.canRetain() && _canAggregate)
```
I'm not sure I understand that `canRetain()` / `canAggregate` logic. If `canRetain()` returns false, this means that RBB's lifecycle is left to the GC, so I don't get why we should aggregate in that case.

gregw replied:
So you are saying that we can assume a non-retainable buffer can be kept by reference and not released, but just GC'd.

Hmmm maybe.    I guess we should say that in the javadoc for canRetain!

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:
```diff
@@ -159,6 +190,370 @@ default void clear()
+            {
+                case 0 -> callback.succeeded();
+                case 1 -> _buffers.get(0).writeTo(sink, last, callback);
+                default -> new IteratingCallback()
```
Don't you need to succeed/fail the callback given as a parameter when this iterating callback finishes up?

gregw replied:
ooops yes.  So I need a better test as well!

gregw committed:
Merge branch 'jetty-12.0.x' into fix/jetty-12/10541/byteBufferAccumulator2

gregw committed:
Don't modify deprecated class

gregw committed:
More updates after review

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:
```diff
@@ -384,27 +426,37 @@ public ByteBuffer getByteBuffer()
                 case 1 -> _buffers.get(0).getByteBuffer();
-                default -> copyBuffer();
+                default ->
+                {
```
The `default` block looks like a dupe of `copy()`, and it leaks the RBB.

gregw replied:
I don't think it leaks because it saves the result in _buffers to be released when the buffer is.  But it could probably use copy to build the buffer....

gregw replied:
I've put in a common impl

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/internal/ContentSourceRetainableByteBuffer.java:
```diff
@@ -0,0 +1,64 @@
+                return;
+            }
+
+            if (Content.Chunk.isFailure(chunk))
```
I'm wondering if we shouldn't take the habit of always adding a config setting to make transient failures ignorable?

gregw committed:
Merge branch 'jetty-12.0.x' into fix/jetty-12/10541/byteBufferAccumulator2

gregw committed:
updated after merge

gregw commented:
@lorban there is a lot that I like in this PR, but I think it needs really careful review and testing.   Lots of aggregation/accumulation needs review and some careful thought if it can be improved.   Feel free to work with me on this one.... no big hurry as it is not needed in next release, but let's not drop it either.

gregw committed:
updated after merge

lorban committed:
replace ChunkAccumulator usages with ContentSourceRetainableByteBuffer and add tests

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix maxSize check, add test, add javadoc

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix Content.Sink.from(OutputStream) and add test

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:
```diff
@@ -159,6 +207,441 @@ default void clear()
+        }
+
+        @Override
+        public RetainableByteBuffer copy()
```
This implementation does not return an independent copy like the contract states.

lorban replied:
Fixed by specifying in the contract that the backing memory is not independent.

sbordet replied:
This should be renamed to slice()!
It does not do any copy!

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:
```diff
@@ -159,6 +207,441 @@ default void clear()
+        @Override
+        public int capacity()
+        {
+            return Math.max(_buffer.capacity(), _maxCapacity);
```
We need to check if the servlet output buffer relies on this code, and if the servlet spec mandates a strict buffer size or if it's lenient in that regard.

gregw committed:
merged copy and getByteBuffer implementation

gregw committed:
handle 0 capacity

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ByteBufferPool.java:
```diff
@@ -127,12 +127,15 @@ private Buffer(ByteBuffer byteBuffer)
      * {@link #release() released} when they are consumed.</p>
+     * @see RetainableByteBuffer.Accumulator
      */
     class Accumulator
```
I think this class should be deprecated as it's been superseded by `RetainableByteBuffer.Accumulator` and we're not using it anymore

lorban reviewed (commented):
I'm still progressing with the review and enhancement of this PR, but I'm also adding some comments for discussion points and/or as a list of things left to do.

lorban commented on jetty-core/jetty-io/src/test/java/org/eclipse/jetty/io/ByteBufferAggregatorTest.java:
Maybe we want to keep the tests around, even if the target class is deprecated?

lorban committed:
do not copy in append()

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
add tests for append(RBB) variant

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
add tests for copy + fix bugs

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
improve javadoc

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix javadoc

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
improve tests

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix releasing of appended RBB

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix leaks and leak detection in tests

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java:
```diff
@@ -561,7 +624,7 @@ static void write(Sink sink, boolean last, String utf8Content, Callback callback
      * {@link #release()} method overridden.</p>
      */
-    public interface Chunk extends Retainable
+    public interface Chunk extends RetainableByteBuffer
```
This is not a good idea.

Chunk is read-only. Now that `RetainableByteBuffer` has `append()` (also a bad idea), it becomes read/write.

gregw replied:
Chunk already has a read/write API as everything is currently done via `getByteBuffer` so we have the possibility of  `chunk.getByteBuffer().put(...)`.    If fact I can't see that we are very rigorous with ensuring the buffer within a chunk is actually read-only, so we should definitely check that regardless of this PR.

The current `ContentSourceByteBuffer` does `accumulator.copyBuffer(chunk.getByteBuffer())`, forcing a data copy when we could just retain.   This PR allows this code to become just: `accumulator.append(chunk)`, with zero copy.       The other alternative is to build a RBB of the chunk passing the `ByteBuffer` and the chunk as `Retainable`, but that is just creating garbage to try to make chunk look read-only, when it will never have read-only API because it exposes the `ByteBuffer` API.

Perhaps we should go the extra step and create a  `RBB.Mutable`, but this is kind of meaningless unless we can completely hide `ByteBuffer` 

See more discussion in the general comments below...

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:
```diff
@@ -152,13 +193,465 @@ default int capacity()
+    /**
+     * An aggregating {@link RetainableByteBuffer} that may grow when content is appended to it.
+     */
+    class Aggregator implements RetainableByteBuffer
```
I don't think this is a good idea.

`Aggregator` should _have-a_ RBB, not _be_ a RBB.

gregw replied:
The whole point of this exercise is to make the RBB API into something that hides the implementation, which might be:
 + a read-only fixed `ByteBuffer` with no space
 + a read/write single `ByteBuffer` with finite space 
 + a read/write single `ByteBuffer` that may grow to a limit
 + a read/write chain of zero copy buffers

Currently we have it that both the appender and the appendee have to agree on the nature of the implementation.    By making RBB able to hide these details, then the appender need not know and only the creator of the appendee has to choose the implementation.   This will not always be perfect, but it is 50% better than before.

See more general discussion below...

gregw replied:
The other reason that Aggregator should be-a RBB is that forces it to have the same append signature.    We have too many similar by different append/put signatures in the code base: ByteBuffer, BufferUtil, appenders, aggregators, etc.     Having a single API on RBB allows many of these needless API differences to be removed.

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:
```diff
@@ -152,13 +193,465 @@ default int capacity()
+     * An accumulating {@link RetainableByteBuffer} that may internally accumulate multiple other
+     * {@link RetainableByteBuffer}s with zero-copy if the {@link #append(RetainableByteBuffer)} API is used
+     */
+    class Accumulator implements RetainableByteBuffer
```
I don't think this is right.

Like `Aggregator` it should _have_ a RBB, not _be_ one.

It breaks Liskov, and it's obvious that it must have its own separate API because it can store more than 2 GiB of buffers, but if you do so, all the methods will throw trying to copy the bytes into a single BB.

gregw replied:
See Aggregator comments.

Also, it is silly to say that since zero-copy accumulation of buffers may allow the total length to exceed 2GB, then we cannot use that efficient implementation for appenders that will not exceed 2GB.

Yes there is an issue of callers try to get a single BB from the API, but the whole point of hiding the implementation is that we should provide suitable convenience methods that will make explicit calls to `getByteBuffer` an anti-pattern. For example we should call `rbb.writeTo(sink, true, callback)` rather than `sink.write(rbb.getByteBuffer(), true, callback)`

Because we do not provide the correct APIs, we have lots of copies happening now when we could avoid them.

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:
```diff
@@ -152,13 +193,465 @@ default int capacity()
+     * @return true if all bytes of the given buffer were copied, false otherwise.
+     * @see BufferUtil#append(ByteBuffer, ByteBuffer)
+     */
+    default boolean append(ByteBuffer bytes)
```
I don't think this is a great idea.
This API is never used by users of RBB.
Its only existence is for users of `Aggregator`.
Not even `Accumulator` uses this API.

As such, I think it should just belong to `Aggregator`, and be removed from the other 2 classes, which have different semantic, and trying to conflate all semantics into one class is not a good idea.

gregw replied:
The idea is that we want to bypass all the crappy legacy API in `BufferUtil` that has many inconsistencies and come up with a new clean API on RBB that is the preferred usage.

So there are many potential uses of this append API, where there is code that uses `BufferUtil.append(ByteBuffer, ByteBuffer)` like:
```java
            BufferUtil.append(_buffer.getByteBuffer(), byteBuffer);
```

which should just be:
```java
            _buffer.append(byteBuffer);
```

I've just started looking and have found at least 3 such places in the code, and I'm sure there are more where the RBB kept as a separate field to the BB.    I will look for more.

gregw replied:
@lorban I know you didn't like the method... and I have removed it, but I still see some code that does:
```java
   BufferUtil.append(normalByteBuffer, retainableByteBuffer.getByteBuffer());
```

So it would be good to be able to write that as:
```java
   retainableByteBuffer.appendTo(normalByteBuffer);
```

Without this method, we cannot avoid the explicit `getByteBuffer()` call, which I think should be an anti-pattern.   Perhaps the first thing to do is to reduce the usages of `normalByteBuffer` to see if we really need `appendTo` (used to be called `writeTo`, which was the wrong name).

sbordet reviewed (commented):
Sorry, I don't like this PR in its current form for the reasons I explained in the comments.

I like the idea of Aggregator and Accumulator, and they should be retained, however with their own minimal APIs, independent of RBB.

Same with Chunk, it must not be retrofitted to RBB, because ideally Chunk is read-only and RBB is read/write, but mostly write.
Different usages, and the fact that now we have a similar APIs is not to be confused with the fact that there is a _is-a_ relationship.
For example, method `clear()` is in RBB but has no use in Chunk.

I think this PR should be vastly simplified: restore everything as it was, and just add Aggregator and Accumulator independently, each with its own independent API.

gregw committed:
Merge branch 'jetty-12.0.x' into fix/jetty-12/10541/byteBufferAccumulator2

gregw committed:
Update from review

Use RBB.append rather than BU.append

gregw commented:
@sbordet @lorban Firstly, I think it is too late for this PR in 12.0.8, let's punt to next cycle but PLEASE give this some consideration early in that cycle so we can have time to land it.   I think there were a few bugs I found/fixed in this, so I might try to factor those out into another PR for 12.0.8.

The idea of this PR is to move towards the model of having a fully functional buffer abstraction that hides details such as: 
 + a read-only fixed `ByteBuffer` with no space
 + a read/write single `ByteBuffer` with finite space 
 + a read/write single `ByteBuffer` that may grow to a limit
 + a read/write chain of zero copy buffers
 + an optimized for fill RBB that keeps the buffer in fill mode mostly.

Ultimately we should be able to deprecate all the inconsistent BufferUtil APIs for append/put etc. and the RBB as the primary API that we use.   It can hide all the flipping details as well various implementations.   There can even be RBB impls optimized for filling, so the buffer is left in fill mode and only flipped when it is used!

gregw committed:
Update from review

Use RBB.append rather than BU.append

sbordet commented:
> The idea of this PR is to move towards the model of having a fully functional buffer abstraction that hides details such as:

I disagree, seems like you want to make RBB to be the jack of all trades, and it is not necessary at all.
We just add a couple of well designed utility classes such as Aggregator and Accumulator and we don't need anything else.

We obviously do not want to allow users to append to a Chunk, and you do not want to add an API to it, only to throw UnsupportedOperationException when they use it.

gregw commented:
> > The idea of this PR is to move towards the model of having a fully functional buffer abstraction that hides details such as:
> 
> I disagree, seems like you want to make RBB to be the jack of all trades, and it is not necessary at all. We just add a couple of well designed utility classes such as Aggregator and Accumulator and we don't need anything else.
> 

This PR actually started as one just replacing the existing Aggregator/Accumulators with better implementations, which I put up for draft review.   However,  @lorban and I worked those a bit and soon realized that they didn't really solve the problems and that there is a better way, hence the current form of this PR. 

We currently have an API impedance between several key APIs:
 + between chunks and RBB.
 + between RBB and BBs
 + between BB and RBBs
 + between RBBs and Content.Sinks

The impedence between these APIs results in:
 + needless copying of data
 + a plethora of intermediate wrappers/converter/utility classes 
   This is resulting in needless copying in many parts of the code.   
 + code readability/maintainability issues.

For example, we have tried to bridge the impedance from BB to RBB with multiple explicit aggregator and accumulator classes.    It has not worked and too often developers write their own new aggregators/accumulator rather than using what is there.  We now have multiple stale implementations.  Even if we put in the effort to add even more new Aggregators/Accumulators that we now say have the "right" API, it doesn't work because they now suffer from the chunk to RBB impedence.

The solution is not to write even more API adaptors/wrappers/convertors etc.  that try to bridge this API impedance.  The solution is to remove the impedance in the first place so that chunks can smoothly flow into RBBs, that can be used for IO to sinks or streams without falling back to BB APIs.     This is achieved by having a good core buffer abstraction that provides all the commonly used methods directly on it, without the need to fall back to BB APIs and then to use BufferUtil or or BB based utility classes.

We already have a good core read side buffer abstraction with RBB that is used in our low level network code.  For some reason, we then break that abstraction with the nearly identical chunk API, only to then struggle to get it back to BB APIs by using various RBB based utility classes.      If we just fleshed out the RBB abstraction a bit, then it can be RBBs all the way through with no impedance, no conversions, no wrappers, no copies (unless wanted).

I agree that it is less than desirable have read/write APIs that throw `UnsupportedOperationExceptions` on the writers when they are read-only.   But many MANY APIs work this way, least not `ByteBuffer` itself as well as `Collection` and many other core APIs.     Thus I think it is a strawman to say that having a read/write API for chunk is a showstopper.  Its primary API is BB and that is already read/write API.       If we really want to have a read-only buffer abstraction, then this PR is the first step towards that as it will reduce the direct usage of the BB API.  Ultimately we can introduce `RBB.Mutable` if that is really what you want, but I think that would be a next step after this PR.  We have to first reduce the direct BB and BufferUtil usages, which this PR does.   We've done this with HttpFields and HttpURI, and it has been worthwhile.   So I'm not opposed to doing it with RBB, but we cannot whilst we are so dependent on BB APIs

gregw commented:
Perhaps in jetty-12.1 we introduce a read-only `Buffer` and read/write `Buffer.Mutable` APIs.  We can then have `@Deprecated interface RetainableByteBuffer extends Buffer.Mutable`

gregw commented:
@sbordet to understand the API impedance I am referring to, consider writing a buffered echo from a sink to a source.

Currently, there are several point of impedance:
 + the sink produces chunks, which can neither be buffered nor written to a sink, and we need to convert to the BB API, which has mostly been done with a copy rather than a reference.  If we are to avoid a copy, then chunk needs to be wrapped as a RBB
 + Our core buffer abstraction is insufficient for aggregating the chunks, so we need to use a utility class with a different API.... but we have about 6 different versions of those... so rather than pick one, we tend to write a new one instead.
 + Once we have buffered/aggregated/accumulated the wrapped chunks in our utility class, then that can't be used with a sink so we have to convert back to BB (maybe causing yet another copy) before we can write.

With this PR, there is no such impedance as our core buffer abstraction has a sufficiently powerful API and chunks are RBBs, so we can accumulate with `buffer.append(chunk)` and we can write back with `buffer.writeTo(response, true, callback)`.     No extra APIs involved, just our core RBB abstaction.

gregw committed:
Merge branch 'jetty-12.0.x' into fix/jetty-12/10541/byteBufferAccumulator2

sbordet commented:
@gregw wrote:

> the sink produces chunks, which can neither be buffered nor written to a sink, and we need to convert to the BB API, which has mostly been done with a copy rather than a reference. If we are to avoid a copy, then chunk needs to be wrapped as a RBB

That is not accurate. Chunk is retainable, and in fact we already have `Content.Source.asRetainableByteBuffer()` that efficiently does what you want. It is currently not used (perhaps a hint that it is not necessary?).
There is no copy, hence no need to wrap the Chunks as RBBs.

> Once we have buffered/aggregated/accumulated the wrapped chunks in our utility class, then that can't be used with a sink so we have to convert back to BB (maybe causing yet another copy) before we can write.

This was analyzed extensively in the past, and we decided it was a bad idea to have more than 1 API in `Sink` to perform writes.
The reason was that `Response` _is-a_ `Sink`, but having more than 1 API made wrapping more complicated, allowed more confusing semantic (can I write using 2 different APIs alternatively, can 1 API call the other, etc.) and did not play well with external applications that rely on the much, much more widely used `ByteBuffer` APIs (e.g. `UTF_8.encode("hello"), etc.).

Maybe there is an impedance, but we need minimal impedance for users, not for us.
Users will use `ByteBuffer` APIs because that's what vastly more common.
`Sink` cannot have multiple write APIs, and the single one must be based on `ByteBuffer`.
We can be efficient and precise at using our own aggregators and accumulators, which surely need to improve, but given the above I fail to see the reason for all the rest yet.

Is there a compelling use case you have found in the code and that I missed that cannot be done without copying?
AFAIK, either we can efficiently retain, or we explicitly want to copy (e.g. Servlet output aggregation).

gregw commented:
> @sbordet wrote:
> Chunk is retainable, and in fact we already have `Content.Source.asRetainableByteBuffer()` that efficiently does what you want. It is currently not used (perhaps a hint that it is not necessary?). There is no copy, hence no need to wrap the Chunks as RBBs.

Looking at the current code base is not a good indication of what is necessary/good.  We have multiple accumulator/aggregators, mostly predating RBB, so of course they are not used with `asRetainableByteBuffer`.    The API impedance is that you need to find an accumulator utility class and then notice that it's API is for RBB and then find another utility class that will wrap the chunk as a RBB.  So that's two conceptual leaps and a needless wrapper class.

As you say, chunk is `Retainable` and only needs to slightest adaption to be used as a RBB.  This PR recognises that similarity and makes them the same, so no adaption is needed.   Currently we have have:
  + `Chunk` is a `Retainable`; and has a `ByteBuffer`; and is part of a sequence of `Chunk`s with a last one.

This PR just changes that to
  + `Chunk` is a `RetainableByteBuffer`; and is part of a sequence of `Chunk`s with a last one.

What is `RetainableByteBuffer` if it not a `Retainable` that has a `ByteBuffer`?   Why isn't a `Chunk` a RBB?    What is the purpose of that adaption? what semantically does it change? 
It is not a read-only adaption, because `Chunk` uses BB which is a read/write API.   If we want `Chunk` to be truly read-only, then we need to abstract away from BB and RBB is the first step in that direction.   Edit: I  note that we already have `Chunk Chunk.asReadOnly()`, so the current code recognizes that sometimes chunks are read/write!

> This was analyzed extensively in the past, and we decided it was a bad idea to have more than 1 API in `Sink` to perform writes...

So we have previously identified that there is a problem, but were unable to find a fix it in the `Sink` API.     This PR has found another way.  It does not change the `Sink` API, but simply makes RBB `Sink` aware so it can efficiently use the existing Sink API.

> Maybe there is an impedance, but we need minimal impedance for users, not for us. Users will use `ByteBuffer` APIs because that's what vastly more common. `Sink` cannot have multiple write APIs, and the single one must be based on `ByteBuffer`. 

Again, this PR is not proposing changing the Sink API and  BB remains as part of the core Sink abstraction.   But that should not mean that every other core abstraction has to devolve down to a single call to some `getByteBuffer()` method, forcing us to copy into one great big buffer.    Instead this PR makes  RBB is `Sink` aware, so can efficiently manage writing to a sink without any explicit usage of BB.   Similarly it make RBB self aware, so RBBs can be appended to RBBs.

> Is there a compelling use case you have found in the code and that I missed that cannot be done without copying? AFAIK, either we can efficiently retain, or we explicitly want to copy (e.g. Servlet output aggregation).

The compelling use-case is that the current code base has:
 + org.eclipse.jetty.io.ChunkAccumulator
 + org.eclipse.jetty.io.ByteBufferPool.Accumulator
 + org.eclipse.jetty.test.client.transport.HttpClientDemandTest.Accumulator
 + org.eclipse.jetty.io.ByteBufferAccumulator
 + org.eclipse.jetty.io.ByteBufferCallbackAccumulator
 + org.eclipse.jetty.io.ByteBufferAggregator
 + org.eclispe.jetty.util.BufferUtil
 + org.eclipse.jetty.util.IO
 + org.eclipse.jetty.io.Content
 
and more... all with APIs doing essentially the same job of copying bytes from a source to a destination, but all with slightly different APIs and often doing unnecessary copies.  Because of API impedance, developers have had to adapt between our core APIs and the result is a mishmash of adaptions and utilities.   This issue was opened as we recognised that problem, but initially thought that all we needed to do was introduce a few "better" utility classes, but this time with "better" APIs.     We tried that, but it didn't fix the problem.  

The problem is that we should not need to adapt between our core abstractions in the first place.  We should not need utilities and wrappers and adaptors to allow our input to efficiently be buffered and for our buffers to be efficiently written to our output.    To quote Elon Musk, "the best part is no part".   Rather than make a better utility/adaptor/wrapper, this PR removes the need for them in the first place.

You appear to be arguing that BBs are the only common API that we can use between Sources and Sinks.     But BB is a bad API.   We've tried working around that with conventions and `BufferUtil`, but you hate `BufferUtil` and constantly argue against its use.  We now have the RBB abstraction that wraps BB, so why not use that to avoid direct usage of BB whenever possible and thus reduce the need for direct interaction with `BufferUtil`?

So some questions for you:
 + why shouldn't our core input abstraction (`Source`) be directly compatible with our core buffer abstraction (RBB)?
 + why shouldn't our core buffer abstraction (RBB) be directly compatible with our core output abstraction (`Sink`)?

gregw commented:
@lorban @sbordet after the call today, I think we agreed that there is a conceptual difference between the read-only chunk and the read-write RBB.... but that currently in the code base there is zero actual semantic difference between the APIs in this regards, and even some mistakes with making chunks look like they could be mutable.

Thus I think the way forward is to make a real API distinction between mutable and immutable retainable buffers.  Chunk would extend the immutable API, whilst the buffer pools would provide the mutable version.

This is a much bigger change and probably a 12.1.0 thing (which the current PR was borderline being anyway).

gregw committed:
Merge branch 'jetty-12.0.x' into fix/jetty-12/10541/byteBufferAccumulator2

gregw committed:
improved javadoc and removed asReadOnly

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java:
```diff
@@ -889,52 +889,6 @@ default Throwable getFailure()
-        /**
-         * @return an immutable version of this Chunk
-         */
-        default Chunk asReadOnly()
```
Unfortunately, I think this method can only be deprecated for removal, as it is currently part of the API.

gregw replied:
Ah good point

gregw replied:
... unless we make this PR a 12.1 thing.... in which case we can deprecate in 12.0.x and remove here.

gregw commented:
Currently on the RBB interface there are only 2 methods that mutate the contents of the buffer:
```java

    /**
     * Copies the contents of the given byte buffer at the end of this buffer.
     * @param bytes the byte buffer to copy from.
     * @return true if all bytes of the given buffer were copied, false otherwise.
     * @throws ReadOnlyBufferException if the buffer is read only
     * @see BufferUtil#append(ByteBuffer, ByteBuffer)
     */
    boolean append(ByteBuffer bytes) throws ReadOnlyBufferException;

    /**
     * Copies the contents of the given retainable byte buffer at the end of this buffer.
     * @param bytes the retainable byte buffer to copy from.
     * @return true if all bytes of the given buffer were copied, false otherwise.
     * @throws ReadOnlyBufferException if the buffer is read only
     * @see BufferUtil#append(ByteBuffer, ByteBuffer)
     */
    boolean append(RetainableByteBuffer bytes) throws ReadOnlyBufferException;
```

Both are clearly marked that they can throw `ReadOnlyBufferException`.

To avoid these two methods, we would need to split RBB into two and do some difficult renaming. Options include :


```java
interface ImmutableRetainableByteBuffer extends Retainable
{
    ByteBuffer getByteBuffer();
}

interface RetainableByteBuffer extends ImmutableRetainableByteBuffer
{
    boolean(append(ImmutableRetainableByteBuffer);
}
```

I  don't like this style of naming, because a mutable `RetainableByteBuffer` is-a `ImmutableRetainableByteBuffer`, but it is not mutable.   So it would be better to have:


```java
interface RetainableByteBuffer extends extends Retainable
{
    ByteBuffer getByteBuffer();
    
    interface Mutable extends RetainableByteBuffer
    {  
        boolean(append(RetainableByteBuffer);
    }
}
```

But now all our ByteBufferPools need to return `RetainableByteBuffer.Mutable` rather than just a `RetainableByteBuffer`.  There will also be many other API changes, so at this point we may as well just go for a shorter name:

```java
interface Buffer extends extends Retainable
{
    ByteBuffer getByteBuffer();
    
    interface Mutable extends Buffer
    {  
        boolean(append(Buffer);
    }
}
```

ByteBuffer pools would operate on `Buffer.Mutable`s and `Chunks` would extend just `Buffer`


It is a lot of disruption just to avoid 2 methods, that are semantic cousins of the `chunk.getByteBuffer().put(...)` APIs.

gregw commented:
I've had a go at the split to a `RetainableByteBuffer.Mutable` in fix/jetty-12/10541/byteBufferAccumulator3
It is not horrible, but there are a lot of changes.... and it is all a but meaningless as it is still possible to create an immutable RBB that contains a mutable BB and to create a RBB.Mutable that contains an immutable BB.    We would need a lot more rigour to ensure that the correct type of buffer is always passed in (and there is no `isImmutable` method on a BB!)

I'll keep going with it, but it is a huge disruption for 2 append methods that are not different to the put methods on BB, that we have lived with forever!

lorban committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into fix/jetty-12/10541/byteBufferAccumulator2

lorban commented:
My feeling on this is that if we want to go down the route of splitting `RetainableByteBuffer` into mutable/immutable interfaces, `getByteBuffer()` should go away, like the following:

```java
interface Buffer extends Retainable
{
    void writeTo(Content.Sink, boolean, Callback);
    void putTo(nio.ByteBuffer);
    
    interface Mutable extends Buffer
    {  
        boolean append(nio.ByteBuffer);
        int fillFrom(EndPoint);
    }
}
```

That would require massive modifications as almost everywhere we rely on `nio.ByteBuffer` (everywhere but `EndPoint` and `Content.Sink`?) we would need to make modifications to use our `Buffer` API instead. I'm not convinced we should go down that route, at least not now. But I also don't think we can improve `RetainableByteBuffer` by splitting it into read-only/read-write while still exposing the `nio.ByteBuffer`.

gregw commented:
> My feeling on this is that if we want to go down the route of splitting `RetainableByteBuffer` into mutable/immutable interfaces, `getByteBuffer()`

I agree... except perhaps not go away entirely, but be rendered rarely used and considered an antipattern.

So all the commonly used `get` methods from BB should be echoed on RBB.  All the commonly used `put` should be reflected in RBB.Mutable, but with name changes to reflect that a flip, then flop is not needed.   i.e. put is renamed to append, and works much the same, except the content in put between the limit and the capacity.   Finally, the commonly used methods on BufferUtil for append etc. should be available on RBB and those methods deprecated.

I actually don't think this will be too much work... but we will see.  I'll keep experimenting in the 3 branch.

gregw commented:
> That would require massive modifications as almost everywhere we rely on `nio.ByteBuffer` (everywhere but `EndPoint` and `Content.Sink`?) we would need to make modifications to use our `Buffer` API instead. 

I have experimented with that in the 3 branch.   There are a lot of good improvements... but then there are think some classes like HttpParser that need massive modifications.

> I'm not convinced we should go down that route, at least not now. 

I agree.  If we had started out that way, then I think it would have been a good way to go.... but now it feels like a jetty-13 thing at best.

> But I also don't think we can improve `RetainableByteBuffer` by splitting it into read-only/read-write while still exposing the `nio.ByteBuffer`

Also agree.   Unless we rigorously address all usages to ensure that we don't have read-only BB inside read-write RBBs (and versa vice), then it is kind of pointless to split the API.    It could be done,  but I just don't see the benefit at this stage.

Meanwhile, this PR, without the split API contains some significant improvements: both in code simplicity and in avoiding needless copies.   The cost is 2 append methods on RBB that clearly indicate in their signatures that the RBB might contain a read-only BB.  Which is no different from the situation we have today, but without the explicit throws clauses in the API.

I don't think we should let this existing issue be a road block to these improvements!

gregw commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:
```diff
@@ -108,17 +113,54 @@ public boolean release()
      */
     ByteBuffer getByteBuffer();
 
+    /**
+     * Creates a copy of this RetainableByteBuffer that is entirely independent, but
+     * backed by the same memory space, i.e.: modifying the ByteBuffer of the original
+     * also modifies the ByteBuffer of the copy and vice-versa.
+     * @return A copy of this RetainableByteBuffer
+     */
+    default RetainableByteBuffer copy()
+    {
+        return new AbstractRetainableByteBuffer(BufferUtil.copy(getByteBuffer()))
```
Either the comment or the implementation is wrong.   Comment implies shallow copy, but implementation is a deep copy!

gregw commented:
@gregw said:
> @lorban said:
> > But I also don't think we can improve `RetainableByteBuffer` by splitting it into read-only/read-write while still exposing the `nio.ByteBuffer`
> Also agree. Unless we rigorously address all usages to ensure that we don't have read-only BB inside read-write RBBs (and versa vice), then it is kind of pointless to split the API. It could be done, but I just don't see the benefit at this stage.

I'm going to contradict myself here a little bit.. and maybe there is a mid ground that will keep @sbordet happy.    If we moved the mutating methods to a `RBB.Mutable` interface, then the aggregators and accumulators could implement that.    For users of normal RBB, we could have a `RBB.asMutable` method that provides only append methods for code that wants to use them rather than call `getByteBuffer()` and then use that API or `BufferUtil`.    Internally, the BB could even be flipped to fill mode for efficiency, so every append need to flip to fill and then flip back to flush.  We'd need some method to say we've finished mutating, so that the original RBB would see the flipped back results.

The cost would be some extra wrappers (or perhaps just a cast down)  and the call to finish mutation... but it might be OK... 

The Chunk API would still have `asMutable()` but it can be overridden to throw if ever called on a chunk.

I'll experiment with that approach.

gregw committed:
Merge branch 'jetty-12.0.x' into fix/jetty-12/10541/byteBufferAccumulator2

gregw commented:
I have essentially split this PR into:
 + #11598 which cleans up Retainable so that Chunk can be is-a RBB
 + #11599 introduce a RBB.Mutable API and clean up some Aggregators and Accumulators (more work needed)
